internal/hcs/vmcompute.go

```diff
@@ -0,0 +1,549 @@
+package hcs
+
+import (
+	gcontext "context"
+	"syscall"
+
+	"github.com/Microsoft/hcsshim/internal/logfields"
+	"github.com/Microsoft/hcsshim/internal/timeout"
+	"github.com/sirupsen/logrus"
+	"go.opencensus.io/trace"
+)
+
+const (
+	stuckSyscallWarning = "Syscall did not complete within operation timeout. This may indicate a platform issue. If it appears to be making no forward progress, obtain the stacks and see if there is a syscall stuck in the platform API for a significant length of time."
+)
+
+func hcsEnumerateComputeSystemsContext(ctx gcontext.Context, query string, computeSystems **uint16, result **uint16) (hr error) {
+	ctx, span := trace.StartSpan(ctx, "HcsEnumerateComputeSystems")
+	defer span.End()
+	span.AddAttributes(trace.StringAttribute("query", query))
+
+	ctx, cancel := gcontext.WithTimeout(ctx, timeout.SyscallWatcher)
```

Comment:
Obviously a lot of boilerplate in common between all these functions. Do you think we should pull this out into a common function instead? Something like:

```go
func execute(ctx context.Context, f func() error, timeout time.Duration) error {
	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()

	done := make(chan error, 1)
	go func() { done <- f() }
	select {
	case <-ctx.Done():
		if ctx.Err() == context.DeadlineExceeded {
			log.G(ctx).WithField(logfields.Timeout, timeout).Warning(stuckSyscallWarning)
		}
		return ctx.Err()
	case err := <-done:
		return err
	}
}
```

Reply:
Oh fine :)